[3.x] [tvOS] Basic tvOS Support

naithar opened the pull request:
Added **basic** tvOS support.  
Implements already closed but not implemented proposal: https://github.com/godotengine/godot-proposals/issues/224

Platform code is based on iOS platform code with minor changes (some iOS stuff isn't available on tvOS). 
Audio interruption was also removed - AppleTV don't seem post interruption notification correctly or in right order, which results in game freezing after application loses focus. 
Keyboard input backed was changed so keyboard input actually works.
Input is handled by controllers. Apple/Siri Remote are also considered controller (micro)

Export code (including plugins) is also based on iOS platform since both of them export into Xcode project. 
But export is also simpler: 
* Launch screen is storyboard based (no need for launch images)
* tvOS project supports empty Apple Team ID
* Icons, Signatures and certificates are handled by Xcode project

Core code has been modified to address tvOS unsupported features - no `fork` and `execvp`.
Rendering code has been modified for tvOS to use same paths/methods as iOS's. 

Mono haven't been ported.
GDNative uses same configuration as iOS platform.

- [x] Exports correctly
- [x] Runs correctly
- [x] Handles input
- [x] Plugins support (example https://github.com/naithar/godot-ios-plugins/tree/feature/tvos-support)
- [x] GDNative support (https://github.com/godotengine/godot_headers/pull/85 https://github.com/godotengine/godot-cpp/pull/512)
- [x] App Store upload

Build container and script PRs: https://github.com/godotengine/build-containers/pull/71, https://github.com/godotengine/godot-build-scripts/pull/21

Since linux-compiled binaries are required to use `use_lto` to pass App Store checks, I'll place macOS built tvOS templates and binaries here: [unofficial tvOS templates](https://s3.eu-central-1.wasabisys.com/naithar/godot/godot_templates/list.html)

naithar committed:
[CI] tvOS workflow

naithar committed:
[Platform] tvOS support

naithar committed:
[tvOS] tvOS export

naithar committed:
[tvOS] GDNative support

touilleMan commented on modules/gdnative/register_types.cpp:
```diff
@@ -218,9 +226,11 @@ void GDNativeExportPlugin::_export_file(const String &p_path, const String &p_ty
 
+			additional_code = additional_code.replace("$callback_registration", callback_registration_string);
+
 			add_ios_cpp_code(additional_code);
```
shouldn't thoses `add_ios_cpp_code` & `add_ios_linker_flags` calls be behing a `if  (p_features.has("iOS"))` condition now ?

naithar replied:
It's a `EditorExportPlugin` method and it's also used for tvOS export to correctly register static GDNative libraries.
Maybe a more clear name should be used, like `add_ios_tvos_cpp_code` or `add_apple_platform_cpp_code`? I'm not really sure.

touilleMan replied:
yup, I think renaming would be the correct thing to do here ;-)

naithar replied:
On the other hand it might be better to just create `add_tvos_<cpp_code|framework|bundle_file>` methods. This way we wouldn't need to worry about breaking compatibility with already written editor plugins.

naithar replied:
I've added `_tvos_` methods to export plugin. This way old plugins would still work and it would also allow to use tvOS specific configuration.

touilleMan replied:
👍

touilleMan reviewed (approved):
⚠️ I only reviewed the gdnative part of this PR (and it seems pretty fine for me 😄 )

paulo-coutinho commented:
When i submit a version to apple (Apple TV) i got this problem. Can anyone check it this problem still?

https://github.com/godotengine/godot/issues/46565

Calinou commented:
@paulo-coutinho Please don't bump other people's pull requests to bring attention to your support question. The current Godot version doesn't support tvOS and that pull request is required for tvOS support to work.

paulo-coutinho commented:
Im not doing it. But Apple reviewer check that something is blocking menu button. After i sad to reviewers that it is working and make a video they review again and continue the process. Im only put here to advice. If we cannot discuss on pull-requests remove the comments, put some rules or make is private, since it is a community/opensource project i can comment where i want to try help the project.

paulo-coutinho commented:
Apple guys sad that menu button need be handled to back from game to stage menu, and after to main menu and after to exit. I don't see in this PR how to handle menu button. I tried start button, select button etc but don't see how to detect it. Without these things no make sense add support for tvos because will not pass on review.

naithar commented:
> I tried start button, select button etc but don't see how to detect it. 

`Menu` button is mapped to `JOY_START`. You can use it to display in-game pause menu. 

https://github.com/godotengine/godot/blob/2472f1bfb3b5b509441e54d4ea63206776a7e33f/platform/tvos/godot_view.mm#L244-L300

> Without these things no make sense add support for tvos because will not pass on review.

tvOS wasn't even reviewed by the core team, and it might not be even merged in the end.
But passing the review has nothing to do with it - as I said, my project passed the review without an issue. And it seems like you don't have in-game pause menu.

paulo-coutinho commented:
Yes, i already tried it:

```
Input.is_joy_button_pressed(0, JOY_START):
    # my code
```

But it don't enter there.

naithar commented:
You might want to try using `_input(event)` or `_unhandled_input(event):` with `if event as InputEventJoypadButton and event.button_index == JOY_START:`. 
The problem is probably in `int joy_id = OSAppleTV::get_singleton()->joy_id_for_name("Remote");` not returning correct id. It might happen if remote has some other name for some reason and I don't think using constant value like `0` is a good idea.

paulo-coutinho commented:
Hi,

For main action button from TV remote controller or PS4 controller on Apple TV this is working:

```
func _process(delta):
	var button_pressed = false
	
	if Input.is_joy_button_pressed(0, 0):
		button_pressed = true

	if button_pressed:
		_on_ButtonStart_pressed()
```

The only problem is with "menu button". I will try detect it.

naithar commented:
As I said, use `_input(event):` or `_unhandled_input(event):` methods to detect `JOY_START`.

naithar committed:
[tvOS] Rework menu/start button processing

Added delay for 'end' press. This should allow to use Input.is_joy_button_pressed with JOY_START.
Added separate tvOS delay settings in project settings.
Added 'tvOS.overrides_menu_button' value. This allows to enable system's default processing of menu button.

paulo-coutinho commented:
Nice. Now we need only a rebuild of it to test. rsrsrsrsrsrs

naithar commented:
> Nice. Now we need only a rebuild of it to test. rsrsrsrsrsrs

Just use `_input(event):` for now. I'll rebuilt only when RC4 is released.

paulo-coutinho commented:
@naithar 

can you compile a new version since rc4 was released?

thanks

naithar commented:
Reverting this PR (as well as one for `master` branch) back to draft for now. 
I'm working on simplifying code sharing between UIKit based platforms - iOS and tvOS. Progress is here: https://github.com/naithar/godot/commit/4c30d177218383e6e1e59237b5af2ef5f9f812d8

It's working, but I'm not happy with the overall code structure/cleanliness for now.

paulo-coutinho commented:
Hi,

Any news when it will be available?

Valeryn4 commented:
Good day! Appeared in yours.
Is it enough just to merge into the 3.4 branch?

Calinou commented:
> Good day! There was a need for tvOS Is it enough just to merge into the 3.4 branch?

This won't be merged in 3.4.x as patch releases are now dedicated to bug fixes only. This can be merged for 3.5 if this is rebased against the latest `3.x` branch and tested again.

paulo-coutinho commented:
> > Good day! There was a need for tvOS Is it enough just to merge into the 3.4 branch?
> 
> This won't be merged in 3.4.x as [patch releases are now dedicated to bug fixes only](https://godotengine.org/article/versioning-change-godot-3x). This can be merged for 3.5 if this is rebased against the latest `3.x` branch and tested again.

@naithar Can you rebase?

Valeryn4 commented:
> > Good day! There was a need for tvOS Is it enough just to merge into the 3.4 branch?
> 
> This won't be merged in 3.4.x as [patch releases are now dedicated to bug fixes only](https://godotengine.org/article/versioning-change-godot-3x). This can be merged for 3.5 if this is rebased against the latest `3.x` branch and tested again.

It was a question to the author of a branch.
I was unable to build a template for the emulator.
Also, the version worries me about the presence of errors in 3.4.beta.
Tomorrow I will try to merge locally with my branch 3.4. I will do this blindly. Without any understanding of the template device.

naithar commented:
> Good day! There was a need for tvOS
> 
> Is it enough just to merge into the 3.4 branch?

I'm planning to work on this PR and some other iOS issues, but I can't really give a time frame for when it will be ready.

You should be able to use this version of PR after some modifications, but I can't name them as I haven't been following development for some time.

Valeryn4 commented:
problem control...
remote control uses axes for dpad
![image](https://user-images.githubusercontent.com/12483173/150725594-061e36ff-9f7a-4314-b76d-7e2ff7de1306.png)

I added this to my dpad block.
Otherwise, it is not possible to make smooth control of the remote control.
The second generation console always uses two vectors.
Therefore, counting the pressure is meaningless. If we want to emulate dpad clicks, we need to read vectors.
It is also necessary to enter deadzone.

Try dpad control in a 2D game. be surprised)

```
InputDefault::JoyAxis jx = {-1, gamepad.dpad.xAxis.value};
OSAppleTV::get_singleton()->joy_axis(joy_id, JOY_ANALOG_LX, jx);

InputDefault::JoyAxis jy = {-1, -gamepad.dpad.yAxis.value}; //inverse y

OSAppleTV::get_singleton()->joy_axis(joy_id, JOY_AXIS_8, jx);
OSAppleTV::get_singleton()->joy_axis(joy_id, JOY_AXIS_9, jx);

//if axis over deadzone -0.5-0.5 - pressed button
if (jx.value < -0.5)
    OSAppleTV::get_singleton()->joy_button(joy_id, JOY_DPAD_LEFT, gamepad.dpad.left.isPressed);
elif (jx.value > 0.5)
    OSAppleTV::get_singleton()->joy_button(joy_id, JOY_DPAD_RIGHT, gamepad.dpad.left.isPressed);
    
```

Valeryn4 commented:
tvOS unsupport NSDocumentDirectory directory.
tvOS prevents writing out of cache or NSUserDefaults

Need use NSUserDefaults (512kb limit)
Need use NSCachesDirectory

naithar commented:
Superseded by https://github.com/godotengine/godot/pull/58976 and https://github.com/godotengine/godot/pull/58977
